src/vsr/replica.zig

```diff
@@ -2585,105 +2520,52 @@ pub fn Replica(
+        /// In this case, it may be impossible for the new primary to repair the missing header as
         /// the rest of the cluster may have already discarded it. We therefore iterate over our
-        /// uncommitted header gaps and compare them with the quorum of do_view_change messages
-        /// received from other replicas, before starting the new view, to discard any that may be
-        /// impossible to repair.
+        /// uncommitted header gaps to discard any that may be impossible to repair.
         ///
         /// For example, if the old primary replicates ops=7,8,9 (all uncommitted) but only op=9 is
         /// prepared on another replica before the old primary crashes, then this function finds a
         /// gap for ops=7,8 and will attempt to discard ops 7,8,9.
-        // TODO To improve availability, potentially call this before the local headers are
-        // repaired during the view change, so that we can participate in nacking headers.
         fn discard_uncommitted_headers(self: *Self) void {
+            assert(self.replica_count > 1);
             assert(self.status == .view_change);
             assert(self.leader_index(self.view) == self.replica);
             assert(self.do_view_change_quorum);
             assert(!self.repair_timeout.ticking);
             assert(self.op >= self.commit_max);
-            assert(self.replica_count > 1);
             assert(self.op - self.commit_max <= config.journal_slot_count);
 
-            const threshold = self.replica_count - self.quorum_replication;
-            if (threshold == 0) {
-                assert(self.replica_count == 2);
-                return;
-            }
+            // While iterating > commit_max does not in itself guarantee that an op is uncommitted
+            // (the old primary may have committed the op shortly before crashing), nevertheless,
+            // if it was committed it would have survived into the new view as a header not a gap.
+
+            // An op cannot be uncommitted if it is definitely outside the pipeline.
+            const op_committed = std.math.max(self.commit_max, self.op -| config.pipeline_max);
 
-            // Iterating > commit_max does not in itself guarantee that the header is uncommitted.
-            // We must also count nacks from the quorum, since the old primary may have committed
-            // another op just before crashing, if there was sufficient quorum. Counting nacks
-            // ensures that the old primary could not possibly have committed the header.
             var op = self.op;
-            while (op > self.commit_max) : (op -= 1) {
+            while (op > op_committed) : (op -= 1) {
                 if (self.journal.header_with_op(op) != null) continue;
 
                 log.debug("{}: discard_uncommitted_headers: op={} gap", .{ self.replica, op });
 
-                var nacks: usize = 0;
-                for (self.do_view_change_from_all_replicas) |received, replica| {
```

Comment:
Yes, much better! Just to clarify that we only remove it here. We still need it for discarding headers where there are no gaps.

However, using it here was never necessary for correctness, since the view change already guarantees the quorum intersection property, i.e. if any header might have committed, then it is guaranteed to survive into the view change. Therefore, if a header does not survive into a view change, if there is a gap across all replicas, then that op (and subsequent ops) could not possibly have committed.

`discard_uncommitted_headers()` is trying to discard headers for which there are gaps, where we are certain the op could not have been committed.

The NACK protocol has a slightly different purpose: discern for sure whether an op for which there is no gap, where the header exists, whether the op is definitely uncommitted, and can therefore be discarded safely if it cannot be repaired timeously.